sympy/polys/rootconditions.py

```diff
@@ -0,0 +1,256 @@
+
+    qs = [p[1] * qi for qi in p]
+    for i in range(1, len(p), 2):
+        qs[i - 1] -= p[i] * p[0]
+    qs = qs[1:]
+
+    cond = p[0] * p[1]
+    cond = _clear_cond_poly(cond, previous_cond, K)
+
+    previous_cond.append(cond)
+
+    return _rec_dup_routh_hurwitz_poly(qs, previous_cond, K)
+
+
+def _clear_cond_poly(cond: PolyElement[Er], previous_cond, K):
+    # Divide out factors known to be positive from previous conditions.
+
+    # There are not controls in that functions on cond and previous_cond,
+    # we assume that at this point, there are no zeroes, ones and negative ones
+    # in previous_cond.
+    # If there are, there will be an infinite while loop.
+    if cond.is_zero:
+        return K(-1)
+
+    for c in previous_cond:
+        if is_totally_ground(c):
+            continue
+
+        cond_quo, r = K.div(cond, c)
+        while not r:
+            cond = cond_quo
+            cond_quo, r = K.div(cond, c)
+
+    # Remove factors of even degree and reduce odd degree factors
+    common_factor, facs_m = cond.sqf_list()
+
+    cond = prod([fac for fac, m in facs_m if m % 2 == 1], K.one)
+    if common_factor < 0:
+        cond = -K.one*cond
+
+    return cond
+
+
+def is_totally_ground(c: PolyElement[Er]) -> bool:
+    """Check if the condition is ground in all subrings"""
+    while c.ring.domain.is_PolynomialRing and c.is_ground:
+        c = c.LC
+    return c.is_ground
+
+
+def _dup_routh_hurwitz_exraw(p: list[Expr]) -> list[Expr]:
+    if all(c.is_Number for c in p):
+        return _dup_routh_hurwitz_exraw_div(p)
+
+    return _dup_routh_hurwitz_exraw_no_div(p)
+
+
+def _dup_routh_hurwitz_exraw_div(p: list[Expr]) -> list[Expr]:
+    """
+    Stability check with divisions, used for numeric cases.
+
+    """
+    if len(p) < 2:
+        return [EXRAW.one]
+
+    if (p[0]*p[1]).is_nonpositive:
+        return [-EXRAW.one]
+
+    if len(p) == 2:
+        return [p[0]*p[1]]
+
+    qs = p.copy()
+    for i in range(1, len(p), 2):
+        qs[i - 1] -= p[i] * p[0] / p[1]
+    qs = qs[1:]
+
+    return [p[0] * p[1]] + _dup_routh_hurwitz_exraw_div(qs)
+
+
+def _dup_routh_hurwitz_exraw_no_div(p: list[Expr]) -> list[Expr]:
+    """
+    Stability check without divisions, used for EXRAW.
+
+    """
+    return _rec_dup_routh_hurwitz_exraw_no_div(p, [])
+
+
+def _rec_dup_routh_hurwitz_exraw_no_div(p: list[Expr],
+                                previous_cond: list[tuple[int, set[Expr]]]) -> list[Expr]:
+    if len(p) < 2:
+        return [EXRAW.one]
+
+    if len(p) == 2:
+        return [_clear_cond_exraw(p[0]*p[1], previous_cond)[0]]
+
+    qs = [p[1] * qi for qi in p]
+    for i in range(1, len(p), 2):
+        qs[i - 1] -= p[i] * p[0]
+    qs = qs[1:]
+
+    cond, previous_cond = _clear_cond_exraw(p[0] * p[1], previous_cond)
+    return [cond] + _rec_dup_routh_hurwitz_exraw_no_div(qs, previous_cond)
+
+
+def _clear_cond_exraw(cond: Expr, previous_cond: list[tuple[int, set]]) -> tuple[Expr, list[tuple[int, set[Expr]]]]:
```

Comment:
There are some long lines like this and also some odd formatting. I suggest just running ruff format on this file.